plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java

```diff
@@ -0,0 +1,216 @@
+ */
+
+package org.opensearch.plugin.insights.core.service;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.common.Nullable;
+import org.opensearch.common.inject.Inject;
+import org.opensearch.common.lifecycle.AbstractLifecycleComponent;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.plugin.insights.core.exporter.QueryInsightsExporter;
+import org.opensearch.plugin.insights.core.exporter.QueryInsightsExporterType;
+import org.opensearch.plugin.insights.rules.model.SearchQueryRecord;
+import org.opensearch.plugin.insights.settings.QueryInsightsSettings;
+import org.opensearch.threadpool.Scheduler;
+import org.opensearch.threadpool.ThreadPool;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+
+/**
+ * Service responsible for gathering, analyzing, storing and exporting data related to
+ * search queries, based on certain dimensions.
+ *
+ * @param <R> The type of record that stores in the service
+ * @param <S> The type of Collection that holds the aggregated data
+ * @param <E> The type of exporter that exports the aggregated and processed data
+ *
+ * @opensearch.internal
+ */
+public abstract class QueryInsightsService<R extends SearchQueryRecord<?>, S extends Collection<R>, E extends QueryInsightsExporter<R>>
+    extends AbstractLifecycleComponent {
+    private static final Logger log = LogManager.getLogger(QueryInsightsService.class);
+    /** enable insight data collection */
+    private boolean enableCollect;
+
+    /** enable insight data export */
+    private boolean enableExport;
+
+    /** The internal thread-safe store that holds the query insight data */
+    @Nullable
+    protected S store;
+
+    /** The exporter that exports the query insight data to certain sink */
+    @Nullable
+    protected E exporter;
+
+    /** The export interval of this exporter, default to 1 day */
+    protected TimeValue exportInterval = QueryInsightsSettings.MIN_EXPORT_INTERVAL;
+
+    /** The internal OpenSearch thread pool that execute async processing and exporting tasks*/
+    protected final ThreadPool threadPool;
+
+    /**
+     * Holds a reference to delayed operation {@link Scheduler.Cancellable} so it can be cancelled when
+     * the service closed concurrently.
+     */
+    protected volatile Scheduler.Cancellable scheduledFuture;
+
+    /**
+     * Create the Query Insights Service object
+     * @param threadPool The OpenSearch thread pool to run async tasks
+     * @param store The in memory store to keep the Query Insights data
+     * @param exporter The optional {@link QueryInsightsExporter} to export the Query Insights data
+     */
+    @Inject
+    public QueryInsightsService(ThreadPool threadPool, @Nullable S store, @Nullable E exporter) {
+        this.threadPool = threadPool;
+        this.store = store;
+        this.exporter = exporter;
```

Comment:
what if there are multiple exporters? Do we need to create a separate service for each of them? 
I'm not sure if we need to E to be defined as generics, you can directly get list of exporters here.